File: apps/api/app/routes/itineraries.py
import time
from typing import Any, Optional
from uuid import UUID

from fastapi import APIRouter, Depends, Header, Query, Request, status
from fastapi.responses import JSONResponse
from pydantic import BaseModel, ConfigDict, Field

from app.core.auth import AuthValidationError, SupabaseJwtValidator
from app.core.itineraries import (
    ItineraryInputCommand,
    ItineraryHistoryQuery,
    ItineraryItemCommand,
    ItineraryService,
    ItineraryStorageError,
    SaveItineraryCommand,
    get_default_itinerary_service,
)
from app.core.profile import AuthenticatedUser


router = APIRouter()
auth_validator = SupabaseJwtValidator()


class ItineraryInputRequest(BaseModel):
    city_id: UUID
    match_id: Optional[UUID] = None
    date: Optional[str] = None
    party_size: int = Field(ge=1, le=20)
    interests: list[str] = Field(default_factory=list, max_length=20)
    pace: str = "balanced"

    model_config = ConfigDict(extra="ignore")

    def to_command(self) -> ItineraryInputCommand:
        return ItineraryInputCommand(
            city_id=self.city_id,
            match_id=self.match_id,
            date=self.date,
            party_size=self.party_size,
            interests=self.interests,
            pace=self.pace,
        )


class ItineraryItemRequest(BaseModel):
    position: int = Field(ge=1)
    item_type: str = Field(max_length=50)
    source_table: Optional[str] = Field(default=None, max_length=80)
    source_id: Optional[UUID] = None
    title: str = Field(max_length=160)
    description: Optional[str] = Field(default=None, max_length=1500)
    starts_at: Optional[str] = None
    ends_at: Optional[str] = None
    area_label: Optional[str] = Field(default=None, max_length=120)
    route_context: dict[str, Any] = Field(default_factory=dict)

    model_config = ConfigDict(extra="ignore")

    def to_command(self) -> ItineraryItemCommand:
        return ItineraryItemCommand(
            position=self.position,
            item_type=self.item_type,
            source_table=self.source_table,
            source_id=self.source_id,
            title=self.title,
            description=self.description,
            starts_at=self.starts_at,
            ends_at=self.ends_at,
            area_label=self.area_label,
            route_context=self.route_context,
        )


class ItineraryPayloadRequest(BaseModel):
    title: str = Field(max_length=120)
    summary: Optional[str] = Field(default=None, max_length=1000)
    items: list[ItineraryItemRequest] = Field(default_factory=list, max_length=50)

    model_config = ConfigDict(extra="ignore")


class SaveItineraryRequest(BaseModel):
    input: ItineraryInputRequest
    itinerary: ItineraryPayloadRequest

    model_config = ConfigDict(extra="ignore")

    def to_command(self) -> SaveItineraryCommand:
        return SaveItineraryCommand(
            input=self.input.to_command(),
            title=self.itinerary.title,
            summary=self.itinerary.summary,
            items=[item.to_command() for item in self.itinerary.items],
        )


def response_meta(request: Request) -> dict[str, Any]:
    started_at = getattr(request.state, "started_at", time.perf_counter())
    return {
        "request_id": request.state.request_id,
        "trace_id": request.state.trace_id,
        "latency_ms": int((time.perf_counter() - started_at) * 1000),
        "retries": request.state.retries,
        "degraded": request.state.degraded,
    }


def unauthorized(request: Request, message: str) -> JSONResponse:
    return JSONResponse(
        status_code=status.HTTP_401_UNAUTHORIZED,
        content={
            "error": {
                "code": "UNAUTHORIZED",
                "message": message,
            },
            "meta": response_meta(request),
        },
    )


def storage_error(request: Request, error: ItineraryStorageError) -> JSONResponse:
    return JSONResponse(
        status_code=error.status_code,
        content={
            "error": {
                "code": "ITINERARY_STORAGE_UNAVAILABLE",
                "message": str(error),
            },
            "meta": response_meta(request),
        },
    )


async def get_current_user(authorization: str = Header(default="")) -> Optional[AuthenticatedUser]:
    if not authorization:
        return None

    scheme, _, token = authorization.partition(" ")
    if scheme != "Bearer" or not token:
        return None

    try:
        user = await auth_validator.validate(token)
        return AuthenticatedUser(user_id=user["user_id"], access_token=user["access_token"])
    except AuthValidationError:
        return None


def get_itinerary_service() -> ItineraryService:
    return get_default_itinerary_service()


@router.post("/itineraries/generate")
async def generate_itinerary(
    request: Request,
    current_user: Optional[AuthenticatedUser] = Depends(get_current_user),
) -> JSONResponse:
    if current_user is None:
        authorization = request.headers.get("authorization", "")
        if authorization:
            return unauthorized(request, "Invalid bearer token.")
        return unauthorized(request, "Missing bearer token.")

    return JSONResponse(
        status_code=status.HTTP_501_NOT_IMPLEMENTED,
        content={
            "error": {
                "code": "ITINERARY_GENERATION_NOT_IMPLEMENTED",
                "message": "Itinerary generation is not implemented until Phase 06.",
                "details": {
                    "phase": "Phase 06",
                    "contract": "POST /itineraries/save accepts generated itinerary payloads for persistence.",
                },
            },
            "meta": response_meta(request),
        },
    )


@router.post("/itineraries/save")
async def save_itinerary(
    payload: SaveItineraryRequest,
    request: Request,
    current_user: Optional[AuthenticatedUser] = Depends(get_current_user),
    itinerary_service: ItineraryService = Depends(get_itinerary_service),
) -> JSONResponse:
    if current_user is None:
        authorization = request.headers.get("authorization", "")
        if authorization:
            return unauthorized(request, "Invalid bearer token.")
        return unauthorized(request, "Missing bearer token.")

    try:
        saved = await itinerary_service.save_itinerary(current_user, payload.to_command())
    except ItineraryStorageError as error:
        return storage_error(request, error)

    return JSONResponse(
        status_code=status.HTTP_200_OK if saved.get("reused") else status.HTTP_201_CREATED,
        content={**saved, "meta": response_meta(request)},
    )


@router.get("/itineraries/me")
async def my_itineraries(
    request: Request,
    current_user: Optional[AuthenticatedUser] = Depends(get_current_user),
    itinerary_service: ItineraryService = Depends(get_itinerary_service),
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=20, ge=1, le=100),
) -> JSONResponse:
    if current_user is None:
        authorization = request.headers.get("authorization", "")
        if authorization:
            return unauthorized(request, "Invalid bearer token.")
        return unauthorized(request, "Missing bearer token.")

    try:
        payload = await itinerary_service.itineraries_for_user(
            current_user,
            ItineraryHistoryQuery(page=page, page_size=page_size),
        )
    except ItineraryStorageError as error:
        return storage_error(request, error)

    return JSONResponse(content={**payload, "meta": response_meta(request)})

File: apps/web/lib/api/itineraries.ts
const API_BASE_URL = process.env.NEXT_PUBLIC_API_BASE_URL ?? "http://localhost:8000";

export type ObservabilityMeta = {
  request_id: string;
  trace_id: string;
  latency_ms: number;
  retries: number;
  degraded: boolean;
};

export type Pagination = {
  page: number;
  page_size: number;
  total_items: number;
  total_pages: number;
  has_next: boolean;
};

export type ItineraryInput = {
  city_id: string;
  match_id: string | null;
  date: string | null;
  party_size: number;
  interests: string[];
  pace: string;
};

export type ItineraryItem = {
  item_id: string;
  position: number;
  item_type: string;
  source_table: string | null;
  source_id: string | null;
  title: string;
  description: string | null;
  starts_at: string | null;
  ends_at: string | null;
  area_label: string | null;
  route_context: Record<string, unknown>;
};

export type Itinerary = {
  itinerary_id: string;
  city_id: string;
  match_id: string | null;
  input_hash: string;
  input: ItineraryInput;
  title: string;
  summary: string | null;
  status: string;
  items: ItineraryItem[];
  created_at: string;
  updated_at: string;
};

export type GenerateItineraryStubResponse = {
  error: {
    code: "ITINERARY_GENERATION_NOT_IMPLEMENTED" | string;
    message: string;
    details?: Record<string, unknown>;
  };
  meta: ObservabilityMeta;
};

export type MyItinerariesResponse = {
  itineraries: Itinerary[];
  pagination: Pagination;
  meta: ObservabilityMeta;
};

export class ItineraryApiError extends Error {
  status: number;
  body: unknown;

  constructor(status: number, body: unknown, fallbackMessage: string) {
    const message =
      typeof body === "object" &&
      body !== null &&
      "error" in body &&
      typeof body.error === "object" &&
      body.error !== null &&
      "message" in body.error &&
      typeof body.error.message === "string"
        ? body.error.message
        : fallbackMessage;
    super(message);
    this.name = "ItineraryApiError";
    this.status = status;
    this.body = body;
  }
}

function authHeaders(accessToken?: string | null): HeadersInit {
  return accessToken ? { Authorization: `Bearer ${accessToken}` } : {};
}

async function parseJsonResponse(response: Response): Promise<unknown> {
  try {
    return await response.json();
  } catch {
    return {};
  }
}

export async function generateItinerary(
  input: ItineraryInput,
  accessToken?: string | null
): Promise<never> {
  const response = await fetch(`${API_BASE_URL}/itineraries/generate`, {
    body: JSON.stringify(input),
    headers: {
      "Content-Type": "application/json",
      ...authHeaders(accessToken),
    },
    method: "POST",
  });
  const body = await parseJsonResponse(response);
  throw new ItineraryApiError(response.status, body, "Unable to generate itinerary.");
}

export async function fetchMyItineraries(
  accessToken?: string | null,
  page = 1,
  pageSize = 20
): Promise<MyItinerariesResponse> {
  const url = new URL(`${API_BASE_URL}/itineraries/me`);
  url.searchParams.set("page", String(page));
  url.searchParams.set("page_size", String(pageSize));

  const response = await fetch(url.toString(), {
    headers: authHeaders(accessToken),
  });
  const body = await parseJsonResponse(response);
  if (!response.ok) {
    throw new ItineraryApiError(response.status, body, "Unable to load itineraries.");
  }
  return body as MyItinerariesResponse;
}
